File: 397.integer-replacement.py
#
# @lc app=leetcode id=397 lang=python3
#
# [397] Integer Replacement
#

# Time:  O(logn)
# Space: O(1)


# @lc code=start
class Solution:
    def integerReplacement(self, n: int) -> int:
        res = 0
        while n != 1:
            if n == 3:
                n -= 1
            elif n & 3 == 3:
                n += 1
            elif n & 3 == 1:
                n -= 1
            else:
                n //= 2
            res += 1
        return res


# @lc code=end

File: 397.integer-replacement.js
/*
 * @lc app=leetcode id=397 lang=javascript
 *
 * [397] Integer Replacement
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var integerReplacement = function (n) {
  let res = 0;
  while (n !== 1) {
    if (n === 3) {
      --n;
    } else if ((n & 3) === 3) {
      ++n;
    } else if ((n & 3) === 1) {
      --n;
    } else {
      n = Math.trunc(n / 2);
    }
    ++res;
  }
  return res;
};
// @lc code=end
